### PR TITLE
Refactored Ubiquity Unit Test to fork mainnet

### LINKED ISSUE #630: Refactor tests remove mockups
from #571 
>None of the current tests are using a fork environment. Instead, tests like for TWAPOracleDollar3Pool that relies on CurveFactory is testing a MockTWAPOracle instead of the actual contract. I don't know how many tests are following this pattern but it means that multiple contracts are currently untested.

Several tests are testing mockup contracts instead of the real ones. They need to be fixed to test the proper contracts and to fork mainnet if necessary.

Offending Tests
`ChefFacet.t.sol`
`CreditNftManagerFacet.t.sol`
`CreditNftRedemptionCalculatorFacet.t.sol`
`CreditRedemptionCalculatorFacet.t.sol`
`ManagerFacet.t.sol`
`StakingFacet.t.sol`
`TwapOracleFacet.t.sol`

Changing `DiamondTestHelper.sol` is also necessary.

### PR DESCRIPTION
**Pull request by molecula451**
#### This PR *Does not* removes or clean Diamonds, diamonds are still being deployed even when forked
diamond

###### Playing with mainnet 
new

#### There is an virtual function added on top of *DiamondTestSetup.sol* for bool checking if mainnet is forked 
override

#### Finally refactored some tests that are playing with mainnet returning mainnet values
pending
pending1

## More tests

There are also other goods tests and non-trivial, like deploying Curve metapool and adding blazing liquidity to it

### CODE REVIEW

**Review by rndquu — COMMENTED** [review]
The thing is that I don't understand the initial [issue](https://github.com/ubiquity/ubiquity-dollar/issues/630). Yes, we should have tests in a forked environment but in my opinion we should separate unit tests (in a none-forked environment) and integrations tests (in a forked environment).

Besides we don't have a production deploy (with the diamond setup) so there is basically nothing to fork. We can basically fork only the UbiquityDollar contract and users' balances.

@molecula451 did his best trying to fix this issue as he understands it

There are minor issues in this PR like using [deprecated abbreviations](https://github.com/ubiquity/ubiquity-dollar/issues/630), [deprecated interfaces](https://github.com/ubiquity/ubiquity-dollar/pull/680/files#diff-aa687b08fc6d6a877feb71980ee002b4684b3b556ed3d9bfc9b4586b2ccd8748R15) or [modified tests](https://github.com/ubiquity/ubiquity-dollar/pull/680/files#diff-692d683764922b0e75e150bce23518983761153e8b2949bc0bee8083b547cf75R111) although the implementation is untouched. 

The main issue with this PR is that it tries to use **old/deprecated** interfaces with a new diamond setup thus making tests even more complicated to understand.

I would:
1. Close this PR as not planned and use it for future reference for integration testing
2. Mark issue as completed
3. Create separate issues for integration testing in a forked environment

@zgorizzo69 what do you think?

**Comment by molecula451** [comment]
> The thing is that I don't understand the initial [issue](https://github.com/ubiquity/ubiquity-dollar/issues/630). Yes, we should have tests in a forked environment but in my opinion we should separate unit tests (in a none-forked environment) and integrations tests (in a forked environment).

that was why I initially asked for brainstorming when started with this one, but to me, it was a good way to get started into the project

> Besides we don't have a production deploy (with the diamond setup) so there is basically nothing to fork. We can basically fork only the UbiquityDollar contract and users' balances.

 Which in turns made no necessary to remove any diamonds testing 

> There are minor issues in this PR like using [deprecated abbreviations](https://github.com/ubiquity/ubiquity-dollar/issues/630), [deprecated interfaces](https://github.com/ubiquity/ubiquity-dollar/pull/680/files#diff-aa687b08fc6d6a877feb71980ee002b4684b3b556ed3d9bfc9b4586b2ccd8748R15) or [modified tests](https://github.com/ubiquity/ubiquity-dollar/pull/680/files#diff-692d683764922b0e75e150bce23518983761153e8b2949bc0bee8083b547cf75R111) although the implementation is untouched.

`

> The main issue with this PR is that it tries to use **old/deprecated** interfaces with a new diamond setup thus making tests even more complicated to understand.

A reminder that not using ABI compatible on mainnet will get us testing `REVERT`

> I would:

> 1. Close this PR as not planned and use it for future reference for integration testing
> 2. Mark issue as completed
> 3. Create separate issues for integration testing in a forked environment

I'd be happy to take the 3 if you guys going to open new issues for alike tasks

**Comment by 0x4007** [comment]
Seems like a solid first contribution. Let me ask @zgorizzo69 on telegram and see what they think about this pull request.

File: packages/contracts/test/diamond/facets/StakingFacet.t.sol [hunk 1]
```diff
@@ -1,27 +1,69 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.19;
 
+interface UGOV {
+    function balanceOf(address account) external view returns (uint256);
+}
+
+interface IManagerMainnet {
```
Comment (zgorizzo69) [inline packages/contracts/test/diamond/facets/StakingFacet.t.sol:8]: not a fan of adding interfaces in a test file like this

**Comment by 0x4007** [comment]
The provided video looks good to me https://github.com/ubiquity/ubiquity-dollar/issues/630#issuecomment-1593771178

File: packages/contracts/test/diamond/facets/StakingFacet.t.sol [hunk 2]
```diff
@@ -1,27 +1,69 @@
...
 import "../../../src/dollar/libraries/Constants.sol";
-import {MockCurveFactory} from "../../../src/dollar/mocks/MockCurveFactory.sol";
-import {MockERC20} from "../../../src/dollar/mocks/MockERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "forge-std/Test.sol";
 
 contract ZeroStateStaking is DiamondSetup {
-    MockERC20 crvToken;
-    uint256 creditNFTLengthBlocks = 100;
+    IERC20 curve3CrvToken = IERC20(0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490);
```
Comment (zgorizzo69) [inline packages/contracts/test/diamond/facets/StakingFacet.t.sol:54]: this are mainnet ETH but what about if we deploy to polygon for instance ? 
A good testing tool would be to have like a generic test setup that you can specialize with addresses from the live network (like adding curve pool address on different mainnet)

File: packages/contracts/test/diamond/facets/ManagerFacet.t.sol [hunk 3]
```diff
@@ -155,47 +169,100 @@ contract ManagerFacetTest is DiamondSetup {
         ];
 
         for (uint256 i = 0; i < mintings.length; ++i) {
-            deal(address(IDollar), mintings[i], 10000e18);
```
Comment (zgorizzo69) [inline packages/contracts/test/diamond/facets/ManagerFacet.t.sol:158]: for test purposes we don't need to impersonate a whale using deal is fine

**Review by zgorizzo69 — CHANGES_REQUESTED** [review]
I am aligned with @rndquu 
and I think we should close the pr 
we could think about a better way to fork env for integration testing and after that indeed create separate integration tests

File: packages/contracts/test/diamond/facets/StakingFacet.t.sol (hunk 1 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/StakingFacet.t.sol:8]: not adding interfaces like this, would clash with other mock up tests that need other interfaces that are not part of mainnet

File: packages/contracts/test/diamond/facets/StakingFacet.t.sol (hunk 2 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/StakingFacet.t.sol:54]: I think that would be a certain entry for a new issue as the first one doesn't mention multi-chain anywhere, so the scope a bit different

File: packages/contracts/test/diamond/facets/ManagerFacet.t.sol (hunk 3 above)
Comment (molecula451) [reply packages/contracts/test/diamond/facets/ManagerFacet.t.sol:158]: impersonating works good when doing forked tests and a much better approach, i think so, but yes

Comment (molecula451) [reply packages/contracts/test/diamond/facets/ManagerFacet.t.sol:158]: this impersonation, creates a pool, add liquidity, transfer to accounts, all within mainnet context & pops outstanding traces

https://www.dropbox.com/s/4mzqgusq27dzsml/metapool.mp4?dl=0

**Comment by molecula451** [comment]
gentlemen it looks like we are 2/2, because by the looks says @pavlovcik is +1 & as much as the same way I worked on the tasked, want the PR to make it, as i know theres value, but maybe @rndquu is 50/50, who owns the last word? @zgorizzo69

**Comment by 0x4007** [comment]
Unfortunately zgorizzo69 is the code owner of this repository and they said

> and I think we should close the pr

Does it make sense to roll in this work as part of the newly created issue #682? 

Unfortunately I don't quite understand all of these details as I am not a Solidity developer.

**Comment by molecula451** [comment]
> Unfortunately zgorizzo69 is the code owner of this repository and they said
> 
> > and I think we should close the pr
> 
> Does it make sense to roll in this work as part of the newly created issue #682?
> 
> Unfortunately I don't quite understand all of these details as I am not a Solidity developer.

this newly opened need more brainstorming as there are points need to be taken, like production deployment mentioned by ndquu, also i notice the label < than a day there, it's not something accomplishable for less than a day

I think this PR mets the requirements for the #630 even tho there may need updates and changes, what about the task there?

**Comment by 0x4007** [comment]
> this newly opened need more brainstorming as there are points need to be taken, like production deployment mentioned by ndquu, also i notice the label < than a day there, it's not something accomplishable for less than a day

> I think this PR mets the requirements for the https://github.com/ubiquity/ubiquity-dollar/issues/630 even tho there may need updates and changes, what about the task there?

@rndquu I'll let you make the decisions here. Feel free to edit the specifications/merge bounties if needed.

**Comment by molecula451** [comment]
> > this newly opened need more brainstorming as there are points need to be taken, like production deployment mentioned by ndquu, also i notice the label < than a day there, it's not something accomplishable for less than a day
> 
> > I think this PR mets the requirements for the #630 even tho there may need updates and changes, what about the task there?
> 
> @rndquu I'll let you make the decisions here. Feel free to edit the specifications/merge bounties if needed.

Let's set the bounty, appreciate so we can start brainstorming the new requirement, I think i can render a layout for the insfrastructure for commenting and feedback

**Comment by molecula451** [comment]
basically I handcrafted this one without enough context on what what's coming, but following the requirements #630  as state. and now this PR provides the context for future value.

Otherwise,

This PR plays in mainnet and there are refactored tests that were not initially playing in mainnet, this PR handle specific ABI to be able to catch mainnet data that otherwise would not be possible without adding it, after all this PR actually pass the tests without interfering with other tests that are not mainnet.

I understand zgorizzo69 but multi-chain was not mentioned till up to this point and in this PR

**Comment by rndquu** [comment]
To sum up

Initially it is our fault because:
1. Issue description is not clear
2. Even if the description had been clear the PR would have added unnecessary complexity (which basically happened)

At the same time the PR requires additional work even if we considered it for merging

Let's:
1. Close this PR without merging
2. Cut bounty reward in half because the work has been done but we can't accept the PR in its current state

@molecula451 Thank you for your contribution, hope you will continue solving other bounties and understand this middle ground decision.

**Pull request closed without merging**